### DISCUSSION TITLE
[Question] Will C# have minor releases again?

### DISCUSSION THREAD

**Discussion by leandromoh** [dotnet/csharplang#3534] [Q&A] [answered]
C# had some minor releases (1.2, 7.1, 7.2, 7.3) and I personally found it amazing because we does not need await the whole package of features (major release) to be done to use greate features and take its benefits. 

I usually look the [implementation status for language features](https://github.com/dotnet/roslyn/blob/master/docs/Language%20Feature%20Status.md) page and currently 8/17 planned features for C# 9 are already merged.

Does the team plans to release minor versions again at some point?

**Comment by HaloFour** [comment]
The C# team does plan on aligning the major releases of the language with the major releases of .NET, and the .NET team is planning on a yearly cadence.  So you should be able to expect a major release per year.  That doesn't answer the question regarding minor releases, but it should mean that you don't have to wait nearly as long between major releases as we have in the past.

**Comment by MgSam** [comment]
I don't have a link but I'm fairly certain Mads has said they are going back to only major versions given it is now a yearly cadence. 

The 7.x minor versions had mixed success, IMHO. While it is nice getting features sooner, the cadence was so short that in one case a mis-feature (`in` keyword) slipped into a release before it was really properly vetted outside of Redmond. In addition, minor version numbers are harder to remember and I've seen first hand it caused confusion among developers. 

It was nice back in the day when .NET and C# shared version numbers for a time as it makes it a lot simpler for devs to understand at a glance. It's a shame that's unlikely to ever be the case again (unless .NET decides to skip a few major version numbers at some point).

**Comment by YairHalberstadt** [comment]
> While it is nice getting features sooner, the cadence was so short that in one case a mis-feature (in keyword) slipped into a release before it was really properly vetted outside of Redmond.

That sounds very much like stating a personal opinion as established fact. I think many users find the `in` feature extremely useful.

**Comment by tannergooding** [comment]
There were some usability concerns raised around `in` after it shipped, such as the hidden copies it can silently introduce.

That one in particular was semi addressed later via the readonly members feature but you still need an external analyzer to warn you about any accidental hidden copies.

Allowing `T(value)` to bind to `T(in value)` has also caused some problems and has caused some API proposals in the framework to prefer `ref` even if the underlying value isnt ever going to be mutated.

**Comment by tannergooding** [comment]
It's still an immensely useful feature, but a longer feedback period might have caught some of those concerns before it had RTM'd.

**Comment by MgSam** [comment]
> > While it is nice getting features sooner, the cadence was so short that in one case a mis-feature (in keyword) slipped into a release before it was really properly vetted outside of Redmond.
> 
> That sounds very much like stating a personal opinion as established fact. I think many users find the `in` feature extremely useful.

It has been widely [panned by the community](https://stackoverflow.com/a/52820464). 

It offers no new expressiveness (its a shortcut for `ref readonly`), most people find it unintuitive (as if you ask 3 people what `in`  does you'll probably get 3 different answers), and it is useful only for rare and narrow use-cases. Finally, it was added to the language during a time when Mads was posting design notes months after the meetings had actually happened, so there was no community feedback before it shipped.

**Comment by 333fred** [comment]
Getting back to the original question:

I won't say never, but with our current yearly release cycle plus the ability to ship previews 6-8 months in advance, the motivation for minor releases has been obviated. The reason they existed was because we wanted to get new bits to customers without having to wait for a new major version of VS. We don't have that requirement anymore: now it's driven by the TFM of your project and we can ship C# previews in RTM versions of VS. So we think we're at a good cadence currently m

**Comment by leandromoh** [comment]
I see! thanks for answers! :) @333fred I confess I do not like to use preview version in production, so would not be greate to release the complete features in a minor C# version?

**Comment by 333fred** [comment, marked as answer]
> so would not be greate to release the complete features in a minor C# version?

Maybe? It would complicate lots of things for those who don't use VS: would we have to make new stable core 3.1 releases? If not, how would developers get the tooling? How do they rationalize their projects to ensure everyone is on the same compiler version? Shipping in lockstep with the SDK significantly simplifies the model here.
